specs/sds_detection_modules.adoc

```diff
@@ -0,0 +1,301 @@
+// vim: tw=79
+
+= SDS detection modules in node_agent
+
+Node agent serves as the component that is responsible for invoking and/or
+executing operations that are part of global Tendrl scope.
+
+Tendrl officially supports specific storage systems (ceph and gluster). However
+Tendrl also aims to be a framework that provides some essential capabilities
+and interfaces so that it is easy to add support for storage systems that are
+not officially supported.
+
+For this Tendrl should be able to auto-detect existing deployments and out of
+the box sane defaults for laying out the storage system and managing them.
+
+== Problem description
+
+Tendrl need to provide a pluggable model using which auto-detection modules
+for different storage systems can be deployed in pre-defined locations and
+Tendrl can load and execute certain generic APIs on these modules to figure the
+underlying pre-existing storage system. This can become the pre-requisite which
+gets executed as part of starting node agent on the storage nodes and helps
+import cluster flow in Tendrl to decide what kind of cluster is getting imported
+. This also would help deciding on deployment of corresponding Tendrl
+integration module for managing the cluster seamlessly.
+
+== Use Cases
+
+This specification covers the import cluster flow in Tendrl and auto detection
+of underlying storage systems on storage nodes.
+
+== Proposed change
+
+We try to introduce a pluggable set of modules (for officially supported storage
+systems from Tendrl) which could deployed in a standard Tendrl defined location
+and would be executed as part of node agent start on nodes. These modules would
+auto detect the underlying storage systems and its related details. These
+details would be tagged with storage nodes so that while import cluster Tendrl
+would be able to decide on deployment of corresponding integration modules on
+the storage systems.
+
+A summary of proposed changes are
+
+* Define a standard path from where all the available storage detection modules
+would be loaded
+
+* While starting the node agent service on the storage nodes, execute these
+plugins and auto detect the storage systems (if any) on underlying storage nodes
+
+* Tag the detected details (if any) with the storage nodes details
+
+=== Alternatives
+
+None
+
+=== Data model impact:
+
+* The entity model for `Node_context` in node-agent would look as below
+
+```
+class NodeContext(EtcdObj):
+    __name__ = 'nodes/%s/Node_context'
+
+    node_id = fields.StrField("node_id")
+    machine_id = fields.StrField("machine_id")
+    fqdn = fields.StrField("fqdn")
+    cluster_id = fields.StrField("cluster_id")
+    sds_pkg_name = fields.StrField("sds_pkg_name")
```

Comment:
Drop both sds_pkg_name and sds_pkg_version. Use only roles. Put up a global namespace for /roles. Each detection module has a hardcoded role that it assigns, eg, ceph/mon, gluster/glusterd etc. The global hierarchy should contain a mapping of roles to all the nodes that belong to that role. Each node itself simply lists the roles in the roles array.

When querying, this would enable the global listing to be queries for all the nodes that are running a ceph mon or glusterd etc. This can be cross referenced against a specific cluster's list of of nodes or against the node listing provided in an import cluster call to verify that all the nodes are being detected.

Also, the sds_pkg_name and sds_pkg_version don't belong. The detection should not be based on rpm packages. It should be based on processes, services, sockets or well known file locations, such as the cli. Versions must be queried by contacting the storage system directly, using the CLI or API call if known. In any case, the objects for the detected cluster are the proper place to put the package information if required, it doesn't belong in the node context.